raster-foundry/raster-foundry
# Allow STAC exports of campaigns

## Overview

This PR eventually will allow STAC exports of campaigns. For now it does the migration / datamodel / db / db test / api work to make it so that that's possible. Next I'll actually start the STAC export adventure.

### Checklist


### Demo

later

### Notes

~Something terrible is happening when trying to write the catalog out at this point. I've added some prints that should show the problem pretty well. When running, you end up with this:~

<details>
```
found the data asset
Asset href: s3://rasterfoundry-development-data-us-east-1/user-uploads/auth0%7C59318a9d2fbbca3e16bcfc92/03733cc6-749c-4000-813c-9772a4d66718/cog.tif
Decoded href: s3://rasterfoundry-development-data-us-east-1/user-uploads/auth0|59318a9d2fbbca3e16bcfc92/03733cc6-749c-4000-813c-9772a4d66718/cog.tif
[ioapp-compute-0] INFO  c.z.h.HikariDataSource - HikariPool-1 - Shutdown initiated...
[ioapp-compute-0] INFO  c.z.h.HikariDataSource - HikariPool-1 - Shutdown completed.
```
</details>

~The `Decoded href` print is right before `IO { GeoTiffRasterSource(decodedHref) } map { rs => ... }`, which has a bunch of other prints inside the `rs => ...` function, but as you can see, the program exits before any of those show up. Worse, with the `attempt`, I'm not getting prints from either branch of the result! I'm very confused.~

~I've tried out `GeoTiffRasterSource(decodedHref)` in a console and it worked fine, so I'm very confused about what's going on, but anyway, something terrible is happening between attempting to create the raster source and `*waves hands*` the rest of the program.~

^^ parents didn't exist 🤦‍♂️

Currently I'm grabbing just the first scene in an annotation project. That's fine for most user-land exports but will be less fine if we need to use campaign exports for annotation projects where we create mosaics for labeling. Similarly, if no scene item can be created for an annotation project, no label item will be created for it. This mirrors current behavior and doesn't introduce any new headaches beyond what's already indicated in raster-foundry/annotate#1027.

Choice of whether to do a campaign or annotation project export is a result of `traverse`-ing the annotation project id and campaign  id fields of the STAC export. Because it's impossible to create a STAC export with both using the API, I didn't bother handling cases where somehow they're both present.

## Testing Instructions

- run migrations
- assemble batch jar
- save this nice export in a file called `export.json`:

```json
{
    "name": "cli-export-test",
    "license": {"license": "proprietary", "url": null},
    "taskStatuses": ["LABELED"],
    "campaignId": "5c725cdf-e8af-4001-a3c4-98211d02f43e"
}
```

- bring up your api server
- get a bearer token using the `dev refresh token` from lastpass -- `echo '{"refresh_token": "the-token"}' | http :9100/api/tokens`
- `export JWT_AUTH_TOKEN=<the bearer token in the response>`
- `cat export.json | http --auth-type=jwt :9100/api/stac | jq .id`
- run your export: `./scripts/console batch "java -cp /opt/raster-foundry/jars/batch-assembly.jar com.rasterfoundry.batch.Main write_stac_catalog <the export id from last step>"`
- it should succeed
- `./scripts/psql`, then `select export_location from exports where id = '<your export id>'`;
- make a directory called `pr-5482-export` and `aws --profile raster-foundry s3 cp <your export location> pr-5482-export/`
- cd to the export dir, unzip your catalog, and `tree` -- `cat foo | jq .links` for the things in your export and confirm they match the structure that you observe

Closes raster-foundry/annotate#974
